### DISCUSSION TITLE
关于类的继承的疑问?

### DISCUSSION THREAD

**Discussion by xingzuhui** [BYJRK/DotNet-Discussions#223] [C# 技术交流]
1.我怎么知道一个类实现了哪些父类和接口呢? 比如button的控件,本类就4000多行代码,加各种继承,真的心累.
2.这些类的方法,属性,是来自哪个类和接口?我看到最终的类,都不知道他有哪些属性和方法,因为很多属性和方法是继承来的.
如果我使用对象浏览器,一来还要运行起来看,二来只能看到最终结果啊.
3.看微软文档很费劲,他是翻译软件给的,不是人工翻译,也不及时.
4.微软的CodeMap老是崩溃,而且我觉得还不是很完美啊.有没有好的vs插件啊

**Comment by xingzuhui** [comment]
codemap用起来.配置步骤:
1.先安装单独部件:
[新GET到的VS功能之Code Map_vs2019 code map插件.pdf](https://github.com/BYJRK/DotNet-Discussions/files/13961218/GET.VS.Code.Map_vs2019.code.map.pdf)

2.解决数据库问题:
[Visual Studio 2022生成解决方案代码图报错“无法连接到指定的数据库” .pdf](https://github.com/BYJRK/DotNet-Discussions/files/13961225/Visual.Studio.2022.pdf)

3.生成效果:
![mvvmtookit 类图](https://github.com/BYJRK/DotNet-Discussions/assets/33748832/34a1a76a-d9ce-424a-9bb3-1a6ba52bc2b8)

**Reply by xingzuhui** [reply]
这个东西很复杂,还不直观

**Reply by ax020913** [reply]
为什么要求企业版vs的才能用呢？搞得用都用不了，www

**Comment by xingzuhui** [comment]
我怎么提出了一个这么基础的问题呢?是因为一个类的元素不一定体现在类里.他有好多来源.比如一个对象的方法,来源有很多:
1.自己写的.
2.继承自父类的,但是自己没写,但是可以使用.
3.重复接口的,自己实现的.
4.接口的扩展方法,注意这里是实现的,继承这个接口就有了,可以直接调用,但是找这个扩展如果的故意让你找不到,可是很难找到的.因为他的类可以随便命名.
5.父类的扩展方法.
6.来自运行时的方法,就是反射的方法,反射是可以在类外调用,定义这个类的信息,当然包含方法.  如股是继承,会不会继承这个反射的方法我没验证.
7.来自程序生成器,编译过程的方法. 就是Roslyn 生成的方法.
8.来自分部类的方法.这个类可以写partial的形式.调用的时候有,你也要找啊,因为方法可能不写在一个文件里,
还有没有补充我也不知道,反正一个方法到底来自那,如果10000行里找不到,还要去七八层级父类,好几个扩展,反射,生成器等地方去找,你知道那感觉吗?

**Reply by xingzuhui** [reply]
这里还都是可以交叉的,比如说继承的方法,是父接口的扩展方法,接口没有;是父类的生成器生成的方法,父类没有;反射生成的方法,本类源代码你也看不到.

**Reply by xingzuhui** [reply]
以后写程序,尽量少反射,性能不行,用特性结合生成器才是未来.现在生成器写法还不方便,等以后改进了,这玩意好用.
拿不到源码,我理解的,Roslyn和反射都是用il代码的树,编程.但是一个是编译,一个是运行时,肯定要编译时啊.
wpf和反射的程序都不支持aot,肯定是他们的函数需要il啊,原生成器就不一样,他第一步需要il,后面可不需要,应该能支持aot的.

**Reply by xingzuhui** [reply]
一个方法是不是实现的接口,祖父接口,在类上也看不出来.

**Reply by xingzuhui** [reply]
还有最近接触到的 类实现了接口的方法,但是类的对象不能调用,他么接口的引用或者转化为接口的引用才能使用,开始我都迷惑好久.
我实现的方法,我自己不能用,多奇葩的设计,但是.net里有,要注意啊,否则你咋看懂代码?

**Reply by xingzuhui** [reply]
哎呀, deepseek解决了我所有的问题,虽然不尽完美,但是解除了我的部分疑惑.如他说,显示接口方法是2个目的,一个是想是实现类隐藏接口方法,二是有2个以上接口方法重名,又要分别实现,是为了解决这个问题.

**Comment by xingzuhui** [comment]
还有一个如下问题:代码
`
namespace App
{
    public class Profile
    {
        public Gender Gender { get; set; }
        public int Age { get; set; }
        public ContactInfo? ContactInfo { get; set; }
    }

    public enum Gender
    {
        Male,
        Female
    }

    public class ContactInfo
    {
        public string? EmailAddress { get; set; }
        public string? PhoneNo { get; set; }
    }

}

using App;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Options;

var configuration = new ConfigurationManager();
configuration.AddJsonFile(path: "profile.json", optional: false, reloadOnChange: true);

new ServiceCollection()
    .AddOptions()
    .Configure<Profile>(configuration)
    .BuildServiceProvider()
    .GetRequiredService<IOptionsMonitor<Profile>>()
    .OnChange(profile =>
    {
        Console.WriteLine($"Gender: {profile.Gender}");
        Console.WriteLine($"Age: {profile.Age}");
        Console.WriteLine($"Email Address: {profile.ContactInfo?.EmailAddress}");
        Console.WriteLine($"Phone No: {profile.ContactInfo?.PhoneNo}\n");
    });
Console.Read();

`

**Reply by xingzuhui** [reply]
问题再这里,.OnChange(profile => ....,怎么这个方法直接就用profile 引用啊. 他也不是linq,直接用变量,我实在不知道也没学过这样的语法啊.
随着学习,我发现以前菜鸟网学的语法,明显不够用.

**Reply by CodingOctocat** [reply]
这里的 `OnChange` 方法的参数是个委托类型，它的具体参数 `profile => { ... }` 是 lambda 表达式，一种简写的委托方式，profile 是随便起的变量名，它本质上相当于把方法 `void XXX(Profile profile)` 传给 `OnChange`，至于为什么 `OnChange` 中的委托的参数类型是 Profile，那是因为 `OnChange` 可能是个泛型方法，上下文有关联到 Profile。

**Comment by CodingOctocat** [comment]
为什么要关心一个类继承了哪些类型？一般常见的自然熟悉，不常见的需要的时候再去查。

**Reply by xingzuhui** [reply]
这是我读源码的时候遇到的问题,当时太多问题了,根本解决不了,就放弃了.
现在深度求索真好,我遇到问题反复详细的问,他也不烦,我用的百度版的.
上面的问题我已经问问了,解决了一部分.真的好老师.

**Reply by ax020913** [reply]
哈哈哈哈deepseek

**Comment by xingzuhui** [comment]
为什么我会有这个疑问的,因为你不了解清楚的,就会重复建类和接口,性能不如系统的,以为理解了,其实啥都不知道.

**Comment by Lj-dog** [comment]
是的 我有时候也很苦恼 vs没有像蓝图blueprint 那样类似于UML的类图，无法查看类与类的关系

**Reply by xingzuhui** [reply]
现在dotnet10 有了附加元素,属性和方法都可以附加了,就更需要慎重,最好是一种作用的扩展做到一个命名空间下,免得混乱.
